lib/plugin.js

```diff
@@ -1,147 +1,127 @@
-    const outputPattern = getOutputPattern(dictionaryPattern || compilerOutputPath);
-    const relativeDictionaryPath = outputPattern.replace(compiler.context, '');
-    const dictionaryRegex = getDictionaryRegex(relativeDictionaryPath);
+    const { moduleEnding, dictionaryPattern, emitFile, keysRoot: relativeKeysRoot } = this.options;
+    const { path: outputPath } = compiler.options.output;
+
+    const keysRoot = rebase(compiler.context, relativeKeysRoot);
+    const dictionaryRegex = getDictionaryRegex(rebase(compiler.context, dictionaryPattern));
 
     compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation, { normalModuleFactory }) => {
-      normalModuleFactory.hooks.afterResolve.tap(PLUGIN_NAME, (module) => {
-        if (module.resource.endsWith(moduleEnding)) {
-          debug('injected module loader for %s', module.resource);
-          // Webpack is unable to accept a loader function directly 🤷‍♂️
-          // A path to a JS file exporting a loader function must be provided.
-          module.loaders.push({
+      normalModuleFactory.hooks.afterResolve.tap(PLUGIN_NAME, (webpackModule) => {
+        if (webpackModule.resource.endsWith(moduleEnding)) {
+          debug('injected module loader for %s', webpackModule.resource);
+          // Webpack only accepts loader names or paths to loader files
+          webpackModule.loaders.push({
             loader: MODULE_LOADER,
-            options: { getId: this.getModuleId, store: this.modules },
+            options: this.modulesOptions,
           });
         }
 
-        // The need for this loader could be avoided by using tokens in the source code
-        // e.g. __WEBPACK_REPLACE_WITH_LOCALE_de-DE__ to be replaced with JSON content.
-        if (dictionaryRegex.test(module.resource)) {
-          debug('injected seed loader for %s', module.resource);
-          module.loaders.push({
-            loader: SEED_LOADER,
-            options: { store: this.seeds },
+        if (dictionaryRegex.test(webpackModule.resource)) {
+          debug('injected replacements loader for %s', webpackModule.resource);
+          webpackModule.loaders.push({
+            loader: REPLACEMENT_LOADER,
+            options: this.replacementsOptions,
           });
         }
       });
 
-      // There is probably a better way to do this, but there is almost no documentation 🤷‍♂️
-      compilation.hooks.optimizeChunkAssets.tapAsync(PLUGIN_NAME, (chunks, callback) => {
-        const replacements = Array.from(this.seeds.keys());
-        if (!replacements.length) return callback();
-
-        debug('attempting to find patterns to replace');
+      compilation.hooks.finishModules.tapPromise(PLUGIN_NAME, async(modules) => {
+        const builtModules = modules.reduce((acc, webpackModule) => {
+          if (!webpackModule.resource || !webpackModule.resource.endsWith(moduleEnding)) return acc;
+          acc.push(getModuleId(keysRoot, moduleEnding, webpackModule.resource));
```

Comment:
reverse if so that you only have one `return acc`

Reply:
and maybe extract into a `util` ?

Reply:
inverse logic is no good, will clean up. not so hot on moving this into utils, we already have `collectModules` which is also basically a 1 reduce function... will think about it